kinesis/kinesis.go

```diff
@@ -194,36 +197,120 @@ func (outputPlugin *OutputPlugin) AddRecord(record map[interface{}]interface{},
 	if err != nil {
 		logrus.Errorf("[kinesis %d] %v\n", outputPlugin.PluginID, err)
 	}
+
+	if retCode == output.FLB_OK {
+		logrus.Debugf("[kinesis %d] Flushed %d logs\n", outputPlugin.PluginID, len(*records))
+	}
+
+	// requestBuf will contain records sendCurrentBatch failed to send
+	*records = requestBuf
 	return retCode
 }
 
+// FlushWithRetries sends the current buffer of log records, with retries
+func (outputPlugin *OutputPlugin) FlushWithRetries(count int, records []*kinesis.PutRecordsRequestEntry) {
+	var retCode, tries int
+
+	currentRetries := outputPlugin.getConcurrentRetries()
+	outputPlugin.addGoroutineCount(1)
+
+	for tries = 0; tries < outputPlugin.concurrencyRetryLimit; tries++ {
+		if currentRetries > 0 {
+			// Wait if other goroutines are retrying, as well as implement a progressive backoff
+			if currentRetries > uint32(outputPlugin.concurrencyRetryLimit) {
+				time.Sleep(time.Duration((1<<uint32(outputPlugin.concurrencyRetryLimit))*100) * time.Millisecond)
+			} else {
+				time.Sleep(time.Duration((1<<currentRetries)*100) * time.Millisecond)
+			}
+		}
+
+		logrus.Debugf("[kinesis %d] Sending (%d) records, currentRetries=(%d)", outputPlugin.PluginID, len(records), currentRetries)
+		retCode = outputPlugin.Flush(&records)
+		if retCode != output.FLB_RETRY {
+			break
+		}
+		currentRetries = outputPlugin.addConcurrentRetries(1)
+		logrus.Infof("[kinesis %d] Going to retry with (%d) records, currentRetries=(%d)", outputPlugin.PluginID, len(records), currentRetries)
+	}
+
+	outputPlugin.addGoroutineCount(-1)
+	if tries > 0 {
+		outputPlugin.addConcurrentRetries(-tries)
+	}
+
+	switch retCode {
+	case output.FLB_ERROR:
+		logrus.Errorf("[kinesis %d] Failed to send (%d) records with error", outputPlugin.PluginID, len(records))
+	case output.FLB_RETRY:
+		logrus.Errorf("[kinesis %d] Failed to send (%d) records after retries %d", outputPlugin.PluginID, len(records), outputPlugin.concurrencyRetryLimit)
+	case output.FLB_OK:
+		logrus.Debugf("[kinesis %d] Flushed %d records\n", outputPlugin.PluginID, count)
+	}
+}
+
+// FlushConcurrent sends the current buffer of log records in a goroutine with retries
+// Returns FLB_OK, FLB_RETRY
+// Will return FLB_RETRY if the limit of concurrency has been reached
+func (outputPlugin *OutputPlugin) FlushConcurrent(count int, records []*kinesis.PutRecordsRequestEntry) int {
+
+	runningGoRoutines := outputPlugin.getGoroutineCount()
+	if runningGoRoutines+1 > int32(outputPlugin.Concurrency) {
+		logrus.Infof("[kinesis %d] flush returning retry, concurrency limit reached (%d)\n", outputPlugin.PluginID, runningGoRoutines)
+		return output.FLB_RETRY
+	}
+
+	curRetries := outputPlugin.getConcurrentRetries()
+	if curRetries > 0 {
+		logrus.Infof("[kinesis %d] flush returning retry, kinesis retries in progress (%d)\n", outputPlugin.PluginID, curRetries)
+		return output.FLB_RETRY
+	}
+
+	go outputPlugin.FlushWithRetries(count, records)
+
+	return output.FLB_OK
```

Comment:
I have a concern with these changes that we're trading off reliability for performance.

My understanding is when we return an `FLB_OK` to the Fluent Bit engine we're acknowledging that the chunk of data Fluent Bit engine handed off was successfully processed. Fluent Bit engine would no longer be able to provide the built-in retry mechanism with that data as it assumes everything has been processed. With these concurrency changes we're hacking around that feature by returning `FLB_OK` earlier than we've processed the data. If concurrency retries are exhausted we're unable to tell the Fluent Bit engine `FLB_RETRY` and we've lost those records.

Reply:
> I have a concern with these changes that we're trading off reliability for performance.

That's true, and that's why this new feature is opt-in.

Reply:
That is a valid concern.  The default behavior is to not enable concurrency, so the Fluent Bit engine retry mechanism will continue to be utilized.

Reply:
I think that's clear to us as we're directly involved in the changes, but for other customers and users I think it would be a surprise to opt in to any feature and end up with data loss. I don't agree with the reliability for performance tradeoff here or the hacky nature of returning `FLB_OK` early, but I can disagree and commit if others think this is the right thing to do. At a minimum we should log a warning message or find a better way to communicate that opting into the feature can result in data loss.

Reply:
Makes sense. Let's explicitly call this out in the docs for this option? 

Zack built this feature because it is necessary for him to reach his throughput needs.  His testing has been pretty thorough and has shown that it does lead to a decent bump in sustained throughput.

Reply:
@zackwine Would you be open to updating the exposed configuration options to include the `experimental_` prefix? I think this will make it clear that stability is not yet **guaranteed** and enable us to more easily pivot in the future if necessary. We'll also release this feature as experimental and update the docs to outline the potential cost associated with using these new fields.

Reply:
@sonofachamp I'll make the changes to prepend `experimental_` to concurrency options.  And add a bit in the documentation about the additional resource usage, and potential for dropped records.

For our use case, the cost of not using the concurrency option is greater.  Enabling concurrency fixes the following 2 major issues for our use case:

- Repeated records when backpressure is received from kinesis (due to the retry logic go fluentbit not being aware a subset of the records were successfully posted).  With the concurrent retry logic if the Kinesis shard throughput is exceeded records will not be duplicated.
- Record throughput goes up about 70-90% with a concurrency value of 3.  This is particularly applicable if the source of records is a file that is rotated (read kubernetes docker logs) where a clock is ticking until the logs are rotated and lost at very high volumes.  Without the concurrency option enabled we will drop more records than with `concurrency=3` in my testing.

Reply:
Thanks for some insight into your use case and results.

We really appreciate your contribution and I'm glad to hear these changes will improve your use case. While I understand for your use case the tradeoff is acceptable, it is not yet clear whether that tradeoff will be acceptable to all users of this feature in the future and the experimental additions will give us some time to get more data before locking it down.

Based on my understanding of the issues you were seeing I believe adding concurrency to the plugin is actually an indirect solution as I don't think the issues mentioned above are unique to only this plugin:
* Configurable options to control record counts or collective payload size could help solve the need to deal with partial record delivery in output plugins
* Concurrency is simply pushing that throughput threshold to missing file rotations a little bit further back

I think both of these should ideally be solved in the contract Fluent Bit provides between the engine and plugins. I'm not sure how feasible this actually is for the Go plugin interface, but hopefully this outlines why I think this concurrency solution is indirect.

Reply:
> Record throughput goes up about 70-90% with a concurrency value of 3. This is particularly applicable if the source of records is a file that is rotated (read kubernetes docker logs) where a clock is ticking until the logs are rotated and lost at very high volumes. Without the concurrency option enabled we will drop more records than with concurrency=3 in my testing.

This is really impressive!!

To add to what Nick said though, we're a little bit uncertain about this change because it kind of fundamentally changes the contract between the plugin and Fluent Bit. Hence the ask to prefix all of the new options with `experimental_`. That leaves open the option to change things in the future, while allowing users to choose this feature after reading the docs and making an informed choice. 

Always returning FLB_OK will break fluent bit's i[nternal monitoring and exported Prometheus metrics](https://docs.fluentbit.io/manual/administration/monitoring)- please add a note in the readme that this feature breaks that as well.

And of course, thank you for all of this effort to try to improve the plugin for everyone!